ophyd/v2/core.py

```diff
@@ -188,14 +190,23 @@ async def connect_children(device: Device, prefix: str, sim: bool):
         async def connect(self, prefix: str = "", sim=False):
             await connect_children(self, prefix + self.prefix, sim)
     """
-    coros = {
-        k: c.connect(prefix, sim)
-        for k, c in device.__dict__.items()
-        if k != "parent" and isinstance(c, Device)
-    }
+
+    coros = {}
+    for name, child_device in get_device_children(device):
+        if isinstance(child_device, DeviceDict):
+            coros.update(child_device.nested_connect(prefix, sim))
+        else:
+            coros.update({name: child_device.connect(prefix, sim)})
+
     await wait_for_connection(**coros)
 
 
+def get_device_children(device: Device) -> Generator[Tuple[str, Device], None, None]:
+    for attr_name, attr in device.__dict__.items():
+        if isinstance(attr, Device):
+            yield f"{attr_name.rstrip('_')}", attr
```

Comment:
I would do the `parent` check here, but not the attribute naming change
```suggestion
        if attr_name != "parent" and isinstance(attr, Device):
            yield attr_name, attr
```